ipa-core/src/protocol/basics/mul/dzkp_malicious.rs

```diff
@@ -80,20 +80,143 @@ impl<'a, F: Field + DZKPCompatibleField<N>, const N: usize>
 
 #[cfg(all(test, unit_test))]
 mod test {
+    use ipa_step_derive::CompactStep;
+
     use crate::{
         error::Error,
         ff::boolean::Boolean,
+        helpers::{Role, Role::H1},
         protocol::{
-            basics::SecureMul,
-            context::{dzkp_validator::DZKPValidator, Context, DZKPContext, UpgradableContext},
+            basics::{
+                mul::{dzkp_malicious::Field, semi_honest::multiplication_protocol, Replicated},
+                SecureMul,
+            },
+            context::{
+                dzkp_field::DZKPCompatibleField,
+                dzkp_validator::{DZKPValidator, Segment},
+                Context, DZKPContext, DZKPUpgradedMaliciousContext, UpgradableContext,
+            },
             RecordId,
         },
         rand::{thread_rng, Rng},
+        secret_sharing::{replicated::semi_honest::AdditiveShare, SharedValueArray, Vectorizable},
         test_fixture::{Reconstruct, Runner, TestWorld},
     };
 
+    /// This function mirrors `zkp_multiply` except that on party cheats.
+    ///
+    /// The cheating party flips `prss_left`
+    /// which causes a flip in `z_left` computed by the cheating party.
+    /// This manipulated `z_left` is then sent to a different helper
+    /// and included in the DZKP batch.
+    pub async fn multiply_with_cheater<'a, F, const N: usize>(
+        ctx: DZKPUpgradedMaliciousContext<'a>,
+        record_id: RecordId,
+        a: &Replicated<F, N>,
+        b: &Replicated<F, N>,
+        prss: &Replicated<F, N>,
+        cheater: Role,
+    ) -> Result<Replicated<F, N>, Error>
+    where
+        F: Field + DZKPCompatibleField<N>,
+    {
+        let mut prss_left = prss.left_arr().clone();
+        if ctx.role() == cheater {
+            prss_left += <<F as Vectorizable<N>>::Array>::from_fn(|_| F::ONE);
+        };
+
+        let z =
+            multiplication_protocol(&ctx, record_id, a, b, &prss_left, prss.right_arr()).await?;
```

Comment:
I find it just adds an extra level of confusion that we achieve this bit flip specifically through manipulating the prss... I would rather just mutate z after computing it.